auth_oidc/models/auth_oauth_provider.py

```diff
@@ -80,3 +84,33 @@ def _parse_id_token(self, id_token, access_token):
 
         res.update(self._map_token_values(res))
         return res
+
+
+class AuthOauthProviderGroupLine(models.Model):
+    _name = 'auth.oauth.provider.group_line'
+
+    provider_id = fields.Many2one('auth.oauth.provider', required=True)
+    group_id = fields.Many2one('res.groups', required=True)
+    expression = fields.Char(required=True, help="Variables: user, token")
+
+    @api.constrains('expression')
+    def _check_expression(self):
+        for this in self:
+            try:
+                this._eval_expression(self.env.user, {})
+            except (AttributeError, KeyError, NameError) as e:
+                raise exceptions.ValidationError('\n'.join(e.args))
+
+    def _eval_expression(self, user, token):
+        self.ensure_one()
+
+        class Defaultdict2(collections.defaultdict):
+            def __init__(self, *args, **kwargs):
+                super().__init__(Defaultdict2, *args, **kwargs)
```

Comment:
Why this Defaultdict2 ?

Reply:
it yields defaultdicts again that in turn yield defaultdicts. This way the user can write expressions in a quite carefree manner like `token['key1']['key2']['key3'] == 'something'` and it doesn't fail when `key1` doesn't exist

Reply:
did I do something stupid here?

Reply:
Nope, just too clever for me :)